File: src/recolor.ts
import { f32, rint } from "./numeric";
import {
  TRANSPARENT,
  type ColorMapping,
  type DecodedGif,
  type FrameResult,
  type RecoloredGif,
  type RgbaFrame,
} from "./types";

/**
 * Replace RGB colors while preserving the original alpha channel.
 *
 * Every mapping is matched against the frame's *original* colors, and a pixel
 * is only changed by the first mapping that matches it. A pixel matches when
 * its largest per-channel difference from the source color is within
 * `tolerance`. With `softness`, pixels in the outer `softness` band of the
 * tolerance range are blended toward the replacement instead of replaced.
 */
export function replaceColors(
  frame: RgbaFrame,
  colorMappings: readonly ColorMapping[],
  tolerance: number,
  softness = 0,
): FrameResult {
  const pixels = new Uint8ClampedArray(frame.data);
  const changedCounts = new Array<number>(colorMappings.length).fill(0);

  const soft = softness > 0 && tolerance > 0;
  const softStart = Math.max(tolerance - softness, 0);
  const softRange = tolerance - softStart;

  for (let i = 0; i < pixels.length; i += 4) {
    const r = pixels[i]!;
    const g = pixels[i + 1]!;
    const b = pixels[i + 2]!;

    for (let index = 0; index < colorMappings.length; index++) {
      const [fromRgb, toRgb] = colorMappings[index]!;
      const colorDistance = Math.max(
        Math.abs(r - fromRgb[0]),
        Math.abs(g - fromRgb[1]),
        Math.abs(b - fromRgb[2]),
      );
      if (colorDistance > tolerance) continue;

      const blendWeight =
        soft && colorDistance > softStart
          ? f32((tolerance - colorDistance) / softRange)
          : 1;

      if (toRgb === TRANSPARENT) {
        // Fade matched pixels toward fully transparent. With softness, the
        // alpha drops proportionally near the tolerance edge; the RGB is
        // left untouched since it is on its way to being invisible.
        pixels[i + 3] = soft
          ? rint(f32(pixels[i + 3]! * f32(1 - blendWeight)))
          : 0;
      } else if (soft) {
        pixels[i] = blend(r, toRgb[0], blendWeight);
        pixels[i + 1] = blend(g, toRgb[1], blendWeight);
        pixels[i + 2] = blend(b, toRgb[2], blendWeight);
      } else {
        pixels[i] = toRgb[0];
        pixels[i + 1] = toRgb[1];
        pixels[i + 2] = toRgb[2];
      }

      changedCounts[index]!++;
      break;
    }
  }

  return {
    frame: { data: pixels, width: frame.width, height: frame.height },
    counts: changedCounts,
  };
}

/** `source + (target - source) * weight` in float32, as numpy computes it. */
function blend(source: number, target: number, weight: number): number {
  return rint(f32(source + f32(f32(target - source) * weight)));
}

export function recolorGif(
  gif: DecodedGif,
  colorMappings: readonly ColorMapping[],
  tolerance: number,
  softness = 0,
): RecoloredGif {
  const totalChangedCounts = new Array<number>(colorMappings.length).fill(0);
  const frames = gif.frames.map((frame) => {
    const result = replaceColors(frame, colorMappings, tolerance, softness);
    result.counts.forEach((count, index) => {
      totalChangedCounts[index]! += count;
    });
    return result.frame;
  });

  return { ...gif, frames, changedCounts: totalChangedCounts };
}

File: crates/gifcc-core/src/recolor.rs
use crate::color::{ColorMapping, Target};

/// Replace RGB colors in an RGBA buffer while preserving the original alpha
/// channel. Returns how many pixels each mapping changed.
///
/// Mappings run in order against the *original* colors, and a pixel is only
/// changed by the first mapping that matches it. A pixel matches when its
/// largest per-channel difference from the source color is within
/// `tolerance`. With `softness`, pixels in the outer `softness` band of the
/// tolerance range are blended toward the replacement instead of replaced.
pub fn replace_colors(
    rgba: &mut [u8],
    mappings: &[ColorMapping],
    tolerance: i32,
    softness: i32,
) -> Vec<u64> {
    let mut changed_counts = vec![0u64; mappings.len()];
    let soft = softness > 0 && tolerance > 0;
    let soft_start = (tolerance - softness).max(0);
    let soft_range = f64::from(tolerance - soft_start);

    for pixel in rgba.as_chunks_mut::<4>().0.iter_mut() {
        let original = [pixel[0], pixel[1], pixel[2]];

        for (index, (from_rgb, to_color)) in mappings.iter().enumerate() {
            let distance = (0..3)
                .map(|c| (i32::from(original[c]) - i32::from(from_rgb[c])).abs())
                .max()
                .unwrap();
            if distance > tolerance {
                continue;
            }

            let weight = if soft && distance > soft_start {
                (f64::from(tolerance - distance) / soft_range) as f32
            } else {
                1.0
            };

            match to_color {
                // Fade matched pixels toward fully transparent. With softness,
                // the alpha drops proportionally near the tolerance edge; the
                // RGB is left untouched since it is on its way to invisible.
                Target::Transparent if soft => {
                    pixel[3] = (f32::from(pixel[3]) * (1.0 - weight)).round_ties_even() as u8;
                }
                Target::Transparent => pixel[3] = 0,
                Target::Color(to_rgb) if soft => {
                    for c in 0..3 {
                        let source = f32::from(original[c]);
                        let target = f32::from(to_rgb[c]);
                        pixel[c] = (source + (target - source) * weight).round_ties_even() as u8;
                    }
                }
                Target::Color(to_rgb) => pixel[..3].copy_from_slice(to_rgb),
            }

            changed_counts[index] += 1;
            break;
        }
    }

    changed_counts
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run(
        pixels: &[[u8; 4]],
        mappings: &[ColorMapping],
        tolerance: i32,
        softness: i32,
    ) -> (Vec<[u8; 4]>, Vec<u64>) {
        let mut rgba: Vec<u8> = pixels.concat();
        let counts = replace_colors(&mut rgba, mappings, tolerance, softness);
        let out = rgba.as_chunks::<4>().0.to_vec();
        (out, counts)
    }

    #[test]
    fn maps_visible_pixels_once_and_preserves_alpha() {
        let (out, counts) = run(
            &[
                [255, 255, 255, 255],
                [250, 250, 250, 255],
                [0, 0, 0, 0],
                [0, 0, 0, 255],
            ],
            &[
                ([255, 255, 255], Target::Color([0, 0, 0])),
                ([0, 0, 0], Target::Color([255, 0, 0])),
            ],
            10,
            0,
        );

        assert_eq!(counts, [2, 2]);
        assert_eq!(
            out,
            [
                [0, 0, 0, 255],
                [0, 0, 0, 255],
                [255, 0, 0, 0],
                [255, 0, 0, 255]
            ]
        );
    }

    #[test]
    fn softens_pixels_near_tolerance_edge() {
        let (out, counts) = run(
            &[
                [255, 255, 255, 255],
                [250, 250, 250, 255],
                [244, 244, 244, 255],
            ],
            &[([255, 255, 255], Target::Color([0, 0, 0]))],
            10,
            10,
        );

        assert_eq!(counts, [2]);
        assert_eq!(
            out,
            [[0, 0, 0, 255], [125, 125, 125, 255], [244, 244, 244, 255]]
        );
    }

    #[test]
    fn makes_matched_pixels_transparent() {
        let (out, counts) = run(
            &[[255, 255, 255, 255], [250, 250, 250, 255], [0, 0, 0, 255]],
            &[([255, 255, 255], Target::Transparent)],
            10,
            0,
        );

        assert_eq!(counts, [2]);
        assert_eq!(
            out,
            [[255, 255, 255, 0], [250, 250, 250, 0], [0, 0, 0, 255]]
        );
    }

    #[test]
    fn fades_alpha_toward_transparent_with_softness() {
        let (out, counts) = run(
            &[
                [255, 255, 255, 255],
                [250, 250, 250, 255],
                [244, 244, 244, 255],
            ],
            &[([255, 255, 255], Target::Transparent)],
            10,
            10,
        );

        assert_eq!(counts, [2]);
        // Exact match goes fully transparent; the edge pixel fades proportionally.
        assert_eq!(
            out,
            [
                [255, 255, 255, 0],
                [250, 250, 250, 128],
                [244, 244, 244, 255]
            ]
        );
    }
}
